### PR TITLE
[Database] StudentScores and TutorScores

### PR DESCRIPTION
**Pull request by sjagla27**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
To make the leaderboards for my bachelor thesis more efficent we want to have the scores of the students closer with the exercises themselves. In the future we also want to get rid of the views that are currently used for the tutor leaderboards and the tutor dashboards.
Also see Issue #2044.

### Description
There will be two new tables added to the database. The student_score table will include the most recent result and score of an exercise for a student. The tutor_score table is an aggregation of the five views that are currently used for the tutor leaderboards and the dashboard.
The appropriate updates for the tables so that the data is correct and up to date will come later.

### Steps for Testing
best tested locally because currently there are no changes that can be seen or are sent to the client
1. Create exercises and participate in them with different users.
2. Assess the submission.
3. Have a look in your db and check if all StudentScores and TutorScores are entered correctly.
4. Change assessments and complain or ask for more feedback and answer those requests.
5. Step 3.

### Test Coverage
- StudentScore: 100%
- StudentScoreResource: 100%
- StudentScoreService: 96%
- TutorScore: 96%
- TutorScoreResource: 100%
- TutorScoreService: 87%

### CODE REVIEW

**Review by MaisiKoleni — COMMENTED** [review]
See comments (and look for more).

File: src/main/java/de/tum/in/www1/artemis/service/TutorScoreService.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            tutorScoreRepository.save(newScore);
+        }
+    }
+
+    /**
+     * Helper method for updating complaints and feedback requests in tutor scores.
+     */
+    private TutorScore addComplaintsAndFeedbackRequests(Result result, TutorScore tutorScore, Exercise exercise) {
+        // add complaints and feedback requests
+        if (result.hasComplaint() == Boolean.TRUE) {
```
[the remark this answers is not in the document]
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/TutorScoreService.java]: Done

File: src/main/java/de/tum/in/www1/artemis/service/TutorScoreService.java [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+            if (complaintOptional.isPresent()) {
+                Complaint complaint = complaintOptional.get();
+
+                // complaint
+                if (complaint.getComplaintType() == ComplaintType.COMPLAINT) {
+                    tutorScore.setAllComplaints(tutorScore.getAllComplaints() + 1);
+                    tutorScore.setComplaintsPoints(tutorScore.getComplaintsPoints() + exercise.getMaxScore());
+
+                    if (complaint.isAccepted() == Boolean.TRUE) {
```
[the remark this answers is not in the document]
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/TutorScoreService.java]: Done

**Comment by krusche** [comment]
Closing this PR because it is open for too long.
Feel free to reopen the PR when it is ready to review.
You might then also simply create a new PR.

**Pull request closed without merging**